File: r/R/message.R
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

#' @include arrow-object.R

#' @title Message class
#'
#' @description
#' `Message` holds an Arrow IPC message, which includes metadata and
#' an optional message body.
#'
#' @usage NULL
#' @format NULL
#' @docType class
#'
#' @section R6 Methods:
#'
#' - `$Equals(other)`: Check if this `Message` is equal to another `Message`
#' - `$body_length()`: Return the length of the message body in bytes
#' - `$Verify()`: Check if the `Message` metadata is valid Flatbuffer format
#'
#' @section Active bindings:
#'
#' - `$type`: The message type
#' - `$metadata`: The message metadata
#' - `$body`: The message body as a [Buffer]
#'
#' @rdname Message
#' @name Message
Message <- R6Class(
  "Message",
  inherit = ArrowObject,
  public = list(
    Equals = function(other, ...) {
      inherits(other, "Message") && ipc___Message__Equals(self, other)
    },
    body_length = function() ipc___Message__body_length(self),
    Verify = function() ipc___Message__Verify(self)
  ),
  active = list(
    type = function() ipc___Message__type(self),
    metadata = function() ipc___Message__metadata(self),
    body = function() ipc___Message__body(self)
  )
)

#' @title MessageReader class
#'
#' @description
#' `MessageReader` reads `Message` objects from an input stream.
#'
#' @usage NULL
#' @format NULL
#' @docType class
#'
#' @section R6 Methods:
#'
#' - `$ReadNextMessage()`: Read the next `Message` from the stream. Returns `NULL` if
#'   there are no more messages.
#'
#' @section Factory:
#'
#' `MessageReader$create()` takes the following argument:
#'
#' * `stream`: An [InputStream] or object coercible to one (e.g., a raw vector)
#'
#' @rdname MessageReader
#' @name MessageReader
#' @export
MessageReader <- R6Class(
  "MessageReader",
  inherit = ArrowObject,
  public = list(
    ReadNextMessage = function() ipc___MessageReader__ReadNextMessage(self)
  )
)

MessageReader$create <- function(stream) {
  if (!inherits(stream, "InputStream")) {
    stream <- BufferReader$create(stream)
  }
  ipc___MessageReader__Open(stream)
}

#' Read a Message from a stream
#'
#' @param stream an InputStream
#'
#' @export
read_message <- function(stream) {
  UseMethod("read_message")
}

#' @export
read_message.default <- function(stream) {
  read_message(BufferReader$create(stream))
}

#' @export
read_message.InputStream <- function(stream) {
  ipc___ReadMessage(stream)
}

#' @export
read_message.MessageReader <- function(stream) {
  stream$ReadNextMessage()
}

File: r/src/message.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "./arrow_types.h"

#include <arrow/ipc/reader.h>
#include <arrow/ipc/writer.h>

// [[arrow::export]]
int64_t ipc___Message__body_length(const std::unique_ptr<arrow::ipc::Message>& message) {
  return message->body_length();
}

// [[arrow::export]]
std::shared_ptr<arrow::Buffer> ipc___Message__metadata(
    const std::unique_ptr<arrow::ipc::Message>& message) {
  return message->metadata();
}

// [[arrow::export]]
std::shared_ptr<arrow::Buffer> ipc___Message__body(
    const std::unique_ptr<arrow::ipc::Message>& message) {
  return message->body();
}

// [[arrow::export]]
int64_t ipc___Message__Verify(const std::unique_ptr<arrow::ipc::Message>& message) {
  return message->Verify();
}

// [[arrow::export]]
arrow::ipc::MessageType ipc___Message__type(
    const std::unique_ptr<arrow::ipc::Message>& message) {
  return message->type();
}

// [[arrow::export]]
bool ipc___Message__Equals(const std::unique_ptr<arrow::ipc::Message>& x,
                           const std::unique_ptr<arrow::ipc::Message>& y) {
  return x->Equals(*y);
}

// [[arrow::export]]
std::shared_ptr<arrow::RecordBatch> ipc___ReadRecordBatch__Message__Schema(
    const std::unique_ptr<arrow::ipc::Message>& message,
    const std::shared_ptr<arrow::Schema>& schema) {
  // TODO: perhaps this should come from the R side
  arrow::ipc::DictionaryMemo memo;
  auto batch = ValueOrStop(arrow::ipc::ReadRecordBatch(
      *message, schema, &memo, arrow::ipc::IpcReadOptions::Defaults()));
  return batch;
}

// [[arrow::export]]
std::shared_ptr<arrow::Schema> ipc___ReadSchema_InputStream(
    const std::shared_ptr<arrow::io::InputStream>& stream) {
  // TODO: promote to function argument
  arrow::ipc::DictionaryMemo memo;
  return ValueOrStop(arrow::ipc::ReadSchema(stream.get(), &memo));
}

// [[arrow::export]]
std::shared_ptr<arrow::Schema> ipc___ReadSchema_Message(
    const std::unique_ptr<arrow::ipc::Message>& message) {
  arrow::ipc::DictionaryMemo empty_memo;
  return ValueOrStop(arrow::ipc::ReadSchema(*message, &empty_memo));
}

//--------- MessageReader

// [[arrow::export]]
std::shared_ptr<arrow::ipc::MessageReader> ipc___MessageReader__Open(
    const std::shared_ptr<arrow::io::InputStream>& stream) {
  return std::shared_ptr<arrow::ipc::MessageReader>(
      arrow::ipc::MessageReader::Open(stream));
}

// [[arrow::export]]
std::shared_ptr<arrow::ipc::Message> ipc___MessageReader__ReadNextMessage(
    const std::unique_ptr<arrow::ipc::MessageReader>& reader) {
  return ValueOrStop(reader->ReadNextMessage());
}

// [[arrow::export]]
std::shared_ptr<arrow::ipc::Message> ipc___ReadMessage(
    const std::shared_ptr<arrow::io::InputStream>& stream) {
  return ValueOrStop(arrow::ipc::ReadMessage(stream.get()));
}
